flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adaptive/allocator/SlotSharingSlotAllocator.java

```diff
@@ -133,7 +138,56 @@ public Optional<VertexParallelismWithSlotSharing> determineParallelism(
         return Optional.of(new VertexParallelismWithSlotSharing(allVertexParallelism, assignments));
     }
 
-    private static Map<JobVertexID, Integer> determineParallelism(
+    /**
+     * Distributes free slots across the slot-sharing groups of the job. Slots are distributed as
+     * evenly as possible while taking the minimum parallelism of contained vertices into account.
+     */
+    private static Map<SlotSharingGroupId, Integer> determineSlotsPerSharingGroup(
+            JobInformation jobInformation, int freeSlots) {
+        int numUnassignedSlots = freeSlots;
+        int numUnassignedSlotSharingGroups = jobInformation.getSlotSharingGroups().size();
+
+        final Map<SlotSharingGroupId, Integer> slotSharingGroupParallelism = new HashMap<>();
+
+        for (Tuple2<SlotSharingGroup, Integer> slotSharingGroup :
+                sortSlotSharingGroupsByUpperParallelism(jobInformation)) {
+            final int groupParallelism =
+                    Math.min(
+                            slotSharingGroup.f1,
+                            numUnassignedSlots / numUnassignedSlotSharingGroups);
+
+            slotSharingGroupParallelism.put(
+                    slotSharingGroup.f0.getSlotSharingGroupId(), groupParallelism);
+
+            numUnassignedSlots -= groupParallelism;
+            numUnassignedSlotSharingGroups--;
+        }
+
+        return slotSharingGroupParallelism;
+    }
+
+    private static List<Tuple2<SlotSharingGroup, Integer>> sortSlotSharingGroupsByUpperParallelism(
```

Comment:
This sorting could also take the slotSharingId into account to achieve a truly stable sorting.

Reply:
nit: Can we reuse `SlotSharingSlotAllocator#getMaxParallelismForSlotSharingGroups` here? (are we maybe missing `Iterable<JobInformation.VertexInformation> getVertices()` in JobInformation interface?

Also, I find "upper" a bit confusing, especially since we call it "max" in a different method. Maybe something like "effective" would be more precise.

Reply:
or "required parallelism"

Reply:
I wanted to avoid "maxParallelism" because it's an overloaded term.
"required" I think goes more in the direction of "minimum" parallelism.
"effective" could work.

Reply:
Meh "effective" would be more appropriate after slots were allocated.
Will use "desired".